// @citation-js/core/lib/plugins/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.config = exports.dict = exports.output = exports.input = exports.list = exports.has = exports.remove = exports.add = void 0;

var input = _interopRequireWildcard(require("./input/"));

exports.input = input;

var output = _interopRequireWildcard(require("./output"));

exports.output = output;

var dict = _interopRequireWildcard(require("./dict"));

exports.dict = dict;

var config = _interopRequireWildcard(require("./config"));

exports.config = config;

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

const registers = {
  input,
  output,
  dict,
  config
};
const indices = {};

const add = (ref, plugins = {}) => {
  let mainIndex = indices[ref] = {};

  if ('config' in plugins) {
    registers.config.add(ref, plugins.config);
    delete plugins.config;
  }

  for (let type in plugins) {
    let typeIndex = mainIndex[type] = {};
    let typePlugins = plugins[type];

    for (let name in typePlugins) {
      let typePlugin = typePlugins[name];
      typeIndex[name] = true;
      registers[type].add(name, typePlugin);
    }
  }
};

exports.add = add;

const remove = ref => {
  let mainIndex = indices[ref];

  for (let type in mainIndex) {
    let typeIndex = mainIndex[type];

    for (let name in typeIndex) {
      registers[type].remove(name);
    }
  }

  delete indices[ref];
};

exports.remove = remove;

const has = ref => ref in indices;

exports.has = has;

const list = () => Object.keys(indices);

exports.list = list;
